Cgen error with mutexes

### Describe the bug

Code: https://vosca.dev/p/c8b88b8807

```v
import sync
import time

struct SafeCounter {
mut:
    mt sync.Mutex = sync.new_mutex()
pub mut:
    value map[string]int
}

fn (mut c SafeCounter) inc(key string) {
    c.mt.@lock()
    defer {
        c.mt.unlock()
    }
    c.value[key]++
}

fn (mut c SafeCounter) value(key string) int {
    c.mt.@lock()
    defer {
        c.mt.unlock()
    }
    return c.value[key]
}

fn main() {
    mut counter := &SafeCounter{}

    for i := 0; i < 5; i++ {
        spawn fn [mut counter] () {
            for j := 0; j < 100; j++ {
                counter.inc('key')
            }
        }()
    }

    time.sleep(1 * time.second)
    println(counter.value('key'))
}

```

### Expected Behavior

No errors as on my macOS machine

### Current Behavior

Output:

```
/tmp/v_60000/../../../../../../box/code.v:28: warning: assignment makes integer from pointer without a cast
/tmp/v_60000/../../../../../../box/code.v:28: warning: cast between pointer and integer of different size
/tmp/v_60000/../../../../../../box/code.v:28: error: field expected
builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

Exited with error status 1
```

### Reproduction Steps

Run code above in playground (Ubuntu 22.04)

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.3 fc4c431.e738d67

### Environment details (OS name and version, etc.)

```
V full version: V 0.3.3 fc4c431.e738d67
OS: linux, Ubuntu 22.04.2 LTS
Processor: 2 cpus, 64bit, little endian, Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz

getwd: /home/pmakhnev/playground
vexe: /home/pmakhnev/v/v
vexe mtime: 2023-04-29 18:00:07

vroot: OK, value: /home/pmakhnev/v
VMODULES: OK, value: /root/.vmodules
VTMP: OK, value: /tmp/v_0

Git version: git version 2.34.1
Git vroot status: weekly.2023.17-12-ge738d671 (1 commit(s) behind V master)
.git/config present: true

CC version: cc (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3
```

Building with `-no-retry-compilation -cc clang-17`

```
==================
   ^~~~~~~~~~~~~~~~~~
/tmp/v_1000/bug.01J0E583B93W38QS5ADRDV7BZE.tmp.c:19170:86: error: incompatible pointer to integer conversion initializing 'int' with an expression of type 'sync__Mutex *' (aka 'struct sync__Mutex *') [-Wint-conversion]
        main__SafeCounter* counter = ((main__SafeCounter*)memdup(&(main__SafeCounter){.mt = sync__new_mutex(),.value = new_map_noscan_value(sizeof(string), sizeof(int), &map_hash_string, &map_eq_string, &map_clone_string, &map_free_string),}, sizeof(main__SafeCounter)));
                                                                                            ^~~~~~~~~~~~~~~~~
1 warning and 1 error generated.
...
==================
```

--- a/vlib/v/checker/struct.v
+++ b/vlib/v/checker/struct.v
@@ -111,9 +111,20 @@ fn (mut c Checker) struct_decl(mut node ast.StructDecl) {
 			if field.has_default_expr {
 				c.expected_type = field.typ
 				field.default_expr_typ = c.expr(mut field.default_expr)
-				if field.typ.is_ptr() && !field.default_expr_typ.is_ptr()
-					&& field.default_expr is ast.StructInit {
-					c.error('reference field must be initialized with reference', field.default_expr.pos())
+				if field.typ.is_ptr() != field.default_expr_typ.is_ptr() {
+					match field.default_expr {
+						ast.CallExpr {
+							err_desc := if field.typ.is_ptr() { 'is' } else { 'is not' }
+							val_desc := if field.default_expr_typ.is_ptr() { 'is' } else { 'is not' }
+							c.error('field ${err_desc} reference but default value ${val_desc} reference',
+								field.default_expr.pos())
+						}
+						ast.StructInit {
+							c.error('reference field must be initialized with reference',
+								field.default_expr.pos())
+						}
+						else {}
+					}
 				}
 
 				// disallow map `mut a = b`

--- a/vlib/v/checker/tests/struct_field_mismatch_ref_err.out
+++ b/vlib/v/checker/tests/struct_field_mismatch_ref_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/struct_field_mismatch_ref_err.vv:6:26: error: field is not reference but default value is reference
+    4 | struct SafeCounter {
+    5 | mut:
+    6 |     mt sync.Mutex = sync.new_mutex()
+      |                          ~~~~~~~~~~~
+    7 | pub mut:
+    8 |     value map[string]int

--- a/vlib/v/checker/tests/struct_field_mismatch_ref_err.vv
+++ b/vlib/v/checker/tests/struct_field_mismatch_ref_err.vv
@@ -1,0 +1,40 @@
+import sync
+import time
+
+struct SafeCounter {
+mut:
+    mt sync.Mutex = sync.new_mutex()
+pub mut:
+    value map[string]int
+}
+
+fn (mut c SafeCounter) inc(key string) {
+    c.mt.@lock()
+    defer {
+        c.mt.unlock()
+    }
+    c.value[key]++
+}
+
+fn (mut c SafeCounter) value(key string) int {
+    c.mt.@lock()
+    defer {
+        c.mt.unlock()
+    }
+    return c.value[key]
+}
+
+fn main() {
+    mut counter := &SafeCounter{}
+
+    for i := 0; i < 5; i++ {
+        spawn fn [mut counter] () {
+            for j := 0; j < 100; j++ {
+                counter.inc('key')
+            }
+        }()
+    }
+
+    time.sleep(1 * time.second)
+    println(counter.value('key'))
+}

